pkg/server/endpoints/authorized_entryfetcher.go

```diff
@@ -22,33 +24,44 @@ var _ api.AuthorizedEntryFetcher = (*AuthorizedEntryFetcherWithEventsBasedCache)
-	lastAttestedNodeEventID      uint
+	mu                                  sync.RWMutex
+	cache                               *authorizedentries.Cache
+	clk                                 clock.Clock
+	log                                 logrus.FieldLogger
+	ds                                  datastore.DataStore
+	cacheReloadInterval                 time.Duration
+	pruneEventsOlderThan                time.Duration
+	sqlTransactionTimeout               time.Duration
+	lastRegistrationEntryEventID        uint
+	lastAttestedNodeEventID             uint
+	missedRegistrationEntryEvents       map[uint]time.Time
+	missedAttestedNodeEvents            map[uint]time.Time
+	receivedFirstRegistrationEntryEvent bool
+	receivedFirstAttestedNodeEvent      bool
 }
 
-func NewAuthorizedEntryFetcherWithEventsBasedCache(ctx context.Context, log logrus.FieldLogger, clk clock.Clock, ds datastore.DataStore, cacheReloadInterval, pruneEventsOlderThan time.Duration) (*AuthorizedEntryFetcherWithEventsBasedCache, error) {
+func NewAuthorizedEntryFetcherWithEventsBasedCache(ctx context.Context, log logrus.FieldLogger, clk clock.Clock, ds datastore.DataStore, cacheReloadInterval, pruneEventsOlderThan, sqlTransactionTimeout time.Duration) (*AuthorizedEntryFetcherWithEventsBasedCache, error) {
 	log.Info("Building event-based in-memory entry cache")
-	cache, lastRegistrationEntryEventID, lastAttestedNodeEventID, err := buildCache(ctx, ds, clk)
+	cache, receivedFirstRegistrationEntryEvent, lastRegistrationEntryEventID, missedRegistrationEntryEvents, receivedFirstAttestedNodeEvent, lastAttestedNodeEventID, missedAttestedNodeEvents, err := buildCache(ctx, ds, clk)
 	if err != nil {
 		return nil, err
 	}
 	log.Info("Completed building event-based in-memory entry cache")
 
 	return &AuthorizedEntryFetcherWithEventsBasedCache{
-		cache:                        cache,
-		clk:                          clk,
-		log:                          log,
-		ds:                           ds,
-		cacheReloadInterval:          cacheReloadInterval,
-		pruneEventsOlderThan:         pruneEventsOlderThan,
-		lastRegistrationEntryEventID: lastRegistrationEntryEventID,
-		lastAttestedNodeEventID:      lastAttestedNodeEventID,
+		cache:                               cache,
+		clk:                                 clk,
+		log:                                 log,
+		ds:                                  ds,
+		cacheReloadInterval:                 cacheReloadInterval,
+		pruneEventsOlderThan:                pruneEventsOlderThan,
+		sqlTransactionTimeout:               sqlTransactionTimeout,
+		lastRegistrationEntryEventID:        lastRegistrationEntryEventID,
+		lastAttestedNodeEventID:             lastAttestedNodeEventID,
+		missedRegistrationEntryEvents:       missedRegistrationEntryEvents,
+		missedAttestedNodeEvents:            missedAttestedNodeEvents,
+		receivedFirstAttestedNodeEvent:      receivedFirstAttestedNodeEvent,
+		receivedFirstRegistrationEntryEvent: receivedFirstRegistrationEntryEvent,
```

Comment:
are receivedFirstRegistrationEntryEvent & receivedFirstAttestedNodeEvent required? 
why is it different to use lastAttestedNodeEventID?
is it to cover eventID == 0? if that is the case you can have a negative value and use the same variable to verify if an ID was really set